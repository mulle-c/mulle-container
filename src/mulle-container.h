//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
