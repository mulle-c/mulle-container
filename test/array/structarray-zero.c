//
//  main.c
//  test-array
//
//  Created by Nat! on 04.11.15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//

#include <mulle-container/mulle-container.h>
#include <mulle-testallocator/mulle-testallocator.h>

#include <stdio.h>


struct demo_struct
{
   char     a;
   int      b;
   char     c;
};


static void  zero_array( void)
{
   struct mulle_structarray   *array;
   struct demo_struct         a;
   struct demo_struct         *p;
   int                        i;

   array = mulle_structarray_create( sizeof( struct demo_struct),
                                     alignof( struct demo_struct),
                                     10,
                                     NULL);


   printf( "%zu", mulle_structarray_get_count( array));
   _mulle_structarray_zero_to_count( array, 20);

   for( i = 0; i < 20; i++)
   {
      a.a = 'Z' - i;
      a.b = i * 1848;
      a.c = i + 'A';

      _mulle_structarray_set( array, i, &a);
   }

   printf( " -> %zu", mulle_structarray_get_count( array));
   _mulle_structarray_zero_to_count( array, 10);

   for( i = 0; i < 10; i++)
   {

      p = _mulle_structarray_get( array, i);
      assert( p->a == 'Z' - i);
      assert( p->b == i * 1848);
      assert( p->c == i + 'A');
   }

   printf( " -> %zu\n", mulle_structarray_get_count( array));

   mulle_structarray_destroy( array);
}


static void  set_count_array( void)
{
   struct mulle_structarray   *array;

   array = mulle_structarray_create( sizeof( struct demo_struct),
                                     alignof( struct demo_struct),
                                     10,
                                     NULL);


   printf( "%zu", mulle_structarray_get_count( array));
   _mulle_structarray_set_count( array, 20);
   printf( " -> %zu", mulle_structarray_get_count( array));
   _mulle_structarray_set_count( array, 10);
   printf( " -> %zu\n", mulle_structarray_get_count( array));

   mulle_structarray_destroy( array);
}


// the mulle_testallocator detects and aborts on leaks
static void  run_test( void (*f)( void), char *name)
{
   mulle_testallocator_reset();
  // printf( "%s\n", name);
   (f)();
   mulle_testallocator_reset();
}


int main(int argc, const char * argv[])
{
   run_test( set_count_array, "set count");
   run_test( zero_array, "zero array");
   return( 0);
}
