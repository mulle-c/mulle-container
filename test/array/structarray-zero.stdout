0 -> 20 -> 10
0 -> 20 -> 10
